### DISCUSSION TITLE
BRouter route duration issue

### DISCUSSION THREAD

**Discussion by MotoUKRider** [devemux86/cruiser#48] [Q&A] [answered]
Within Cruiser app, using BRouter - the estimated time for a route using one of my (now famous) GPX files is grossly over-estimated at 172mi, 7:37 but reverts to 185mi, 3:42 using GraphHopper-offline (there is a minor alteration to the route between algorithms, but that's a *big* time difference). More importantly, BRouter on Kurviger v2 app gives me 172mi, 4:48. However, with the longer BRouter calculation, if I then start navigating, I get a much more realistic (ie: nearer the 4hr mark) ETA.

BaseCamp says 173mi, 3:03 and ITNConv (using a Here map) says 173mi, 3:54.  I accept some variations between algorithms, but 3-4hrs is pushing it :(   GraphHopper is a bit of a non-starter for me mainly because it won't take me over a cattle-grid - and I already had to remove "gate" from the BRouter profile to make it work for me.

This is using "Route" as the timing source, and if I use "Speed" then adjust the average speed slider, I get a varied duration - so I'm confident that there is an issue with the current app during the loading of a GPX - more so because if I manually tap+plot the same route within the app, I get the shorter more realistic route time.

Hopefully this can be replicated by trying the following GPX in old v2 app and current build with BRouter, and I'm not just having a bad day :)

[east-coast-route.zip](https://github.com/devemux86/cruiser/files/10791268/east-coast-route.zip)

**Comment by devemux86** [comment]
Thanks for the report!

You probably use [BRouter 1.6.4 (Beta)](https://github.com/devemux86/cruiser/discussions/15)?

> More importantly, BRouter on Kurviger v2 app gives me 172mi, 4:48.

K1 / K2 can only communicate with BRouter with the old method.

BRouter 1.6.4 has since advanced with new features, which can be _recognized / used by Cruiser_.
So it looks like BRouter's new time calculations are not accurate?

> if I manually tap+plot the same route within the app, I get the shorter more realistic route time.

What do you mean, using BRouter (offline)?
If we move slightly a waypoint, we still get the same route duration.

In any case I will check the code again to see what happens.

**Reply by MotoUKRider** [reply]
Well, wouldn't you know that the only specific thing I didn't mention was the *beta* of BRouter that I was using :)

It wouldn't let me re-install the "current" release of BRouter without an uninstall, so I had to re-download my offline routing data before trying this....  and it reports a more expected 172mi, 4:43.  For info, there is still the issue of it making creative adjustments to my GPX (deleted waypoints etc as per other thread) so that's *not* a BRouter beta issue as far as I can tell.  It did take me a while to find a GPX route to use in this thread that wasn't being creatively re-worked before being able to compare against different routing algorithms and the older v2 app.

Manual tap+plot meant that I re-created the GPX route by hand within the app on my phone.  I had a different, more personal route that has 3 or 4 routing points and it also had a vast duration difference between the GPX load and manually creating it.  That was with the beta BRouter.

The main thing is that, after rolling-back BRouter, I'm seeing what I *should* see - but you've obviously found a coding issue beyond my investigations.... Bon appetit ;)

**Reply by devemux86** [reply]
Thanks for the testing!

> For info, there is still the issue of it making creative adjustments to my GPX

Please wait for Android Cruiser 3.0.10 with the import improvements.
(probably today)

Desktop Cruiser 3.0.10 (Beta) is already available and should work properly: https://github.com/devemux86/cruiser/discussions/39#discussioncomment-5055834

> it reports a more expected 172mi, 4:43

This is because Cruiser sees the older BRouter and uses older code.
But we need to see the future and improve the BRouter 1.6.4 case.

**Comment by devemux86** [comment]
@MotoUKRider thanks for the very useful report!

It will work properly in the next Cruiser update:

<img src="https://user-images.githubusercontent.com/3484020/220362119-3fe0eab7-f853-49c0-9c7d-c6127cd0db77.jpg" width="50%"/>

**Comment by devemux86** [comment, marked as answer]
It should work properly with [Cruiser 3.0.10](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Comment by MotoUKRider** [comment]
Digging this one up, again!

I'm convinced there's either something not quite right here or BRouter has a hidden setting that only applies to routes with several Shaping+Via-Points.

I'm seeing 4:40 (BRouter) and 3:27 (GraphHopper) for the planned GPX route (similar to above sample).  Yet, if I remove all the intermediate points and simply make it route Start-to-End, the route stays the same and the mileage/timing also stays more/less the same. HEREMaps says 3:54, GoogleMaps 4:09.  The speed settings in BRouter's `car-fast` profile look OK to me.

If GraphHopper wasn't being a pain with long basic routes (node error) then I could test this much easier :)

I would assume that if the speed-limit data for the roads is applied the same to both Routing engines, that the route times should be very close to each other?!

**Reply by devemux86** [reply]
@MotoUKRider thanks for keep testing!

I tested the above GPX with default settings using BRouter (offline) + GraphHopper (offline) + BRouter website.
The results are presented below. Indeed BRouter and GraphHopper return the same duration with 2 waypoints.

Important: _BRouter (offline) and BRouter website return the same duration for the GPX with all waypoints._

It seems that the algorithms of each router work differently even in the calculation of the route duration.

cc @afischerdev

|BRouter (offline)|
|-|
|<img src="https://github.com/devemux86/cruiser/assets/3484020/6c0ad7a2-bee3-4b42-98f8-9a8cd6effddf" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/2a1120c7-4ba4-45d5-8560-5d03136f6802" width="50%"/>|

|GraphHopper (offline)|
|-|
|<img src="https://github.com/devemux86/cruiser/assets/3484020/dbe4c32c-4edd-4313-b0a3-1c47ce077bdf" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/bc31e63e-83a2-4228-ae43-f946a7e2b7a2" width="50%"/>|

|BRouter website|
|-|
|![Screenshot 2023-05-25 at 10-29-16 BRouter web client](https://github.com/devemux86/cruiser/assets/3484020/c30e6c40-7f8b-4870-a2a3-d4a76e378341)|

**Reply by MotoUKRider** [reply]
Thank you for confirming my sanity :)
Out of the box, I'd expect any routing to be similar unless I'd gone "silly" with various settings, avoidances or set my road speeds to walking-pace.  Fun!

**Reply by devemux86** [reply]
You can also use the "Settings | Routing | Speeds | Route duration"
that calculates the route duration with your selected average speed.

Then the route duration should be the same and according to your personal driving style!

<img width="33%" src="https://github.com/devemux86/cruiser/assets/3484020/e4b3e7eb-9fb7-442f-a285-84237fc4210b"/>

**Reply by MotoUKRider** [reply]
Yes, I had thought of that.  I can see that setting "Car" to 70mph and "Motorcycle" to 50mph may be an option and then using either profile depending upon how I am travelling (eg: car journey for work or fun twisty ride on the bike).  The significance of my above observations (and your useful screenshots) is that it offers a misrepresentation of route times so, when planning a bike route, I'm thinking "I should cut the distance a little as that's a long time in the saddle" :(

Looking at this further - I *know* that the time left on my route will be based on the speed I'm at *on that road* but also assume the posted speed-limit for future roads *unless Route/Speed config is set*. _So, if my route is on a 60mph road for 60miles, my calculated route should come out to a round 1hour, yes? If I select "Speed" and reduce it to 30mph, then I'd expect the calculation to offer 2hours._

The routing engines should assume the speed-limits for each road (there's no traffic service to worry about) and only vary it based on *current speed + future road speed-limits* once on the move or navigating+stationary.

**Reply by devemux86** [reply]
> The routing engines should assume the speed-limits for each road

Exactly, that's how they work. They use the speed tags of the map data.
If the speed tags are empty, they use the road types with preset speeds.

But the speed tags may be used differently by each router.
And the preset speeds / road type may also differ in each router.

(also Google Maps may apply traffic on top)
